### PR TITLE
Drop logs

### PR DESCRIPTION
**Pull request by kfird214**

### CODE REVIEW

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 1]
```diff
@@ -201,6 +445,7 @@ public async ValueTask DisposeAsync()
                 channel.Writer.Complete();
                 cancellationTokenSource.Cancel();
                 await writeLoop.ConfigureAwait(false);
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs:447]: change to `Task.WhenAll`

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 2]
```diff
@@ -1,62 +1,212 @@
...
+        readonly Stream                  stream;
+        readonly Channel<IZLoggerEntry>  channel;
+        readonly Task                    writeLoop;
+        readonly Task                    summaryWriteLoop;
+        readonly ZLoggerOptions          options;
         readonly CancellationTokenSource cancellationTokenSource;
 
         public AsyncStreamLineMessageWriter(Stream stream, ZLoggerOptions options)
         {
-            this.newLine = Encoding.UTF8.GetBytes(Environment.NewLine);
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: stop committing these changes, it's really annoying. please revert all the lines you didn't actually change

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 3]
```diff
@@ -1,62 +1,212 @@
...
             });
-            this.writeLoop = Task.Run(WriteLoop);
+
+            this.writeLoop        = Task.Run(WriteLoop);
+            this.summaryWriteLoop = Task.Run(SummaryWriteLoop);
         }
 
+        private const bool ENABLE_SPAM_DROPPER  = true;
+        private const bool DEBUG_SPAM           = false;
+        private const uint LIMIT_IN             = 3;
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: we need to fix those limits before merging

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 4]
```diff
@@ -1,62 +1,212 @@
...
+        private readonly ConcurrentQueue<DateTimeOffset> postTimesTempQ              = new();
+        private readonly ArrayBufferWriter<byte>         bufferWriterForDebugSpamMsg = new();
+        
+        private static readonly object lockObject = new();
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Post(IZLoggerEntry log)
         {
-            channel.Writer.TryWrite(log);
+            // lock (lockObject)
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: no thanks

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 5]
```diff
@@ -1,62 +1,212 @@
...
+                        lock (lockObject) channel.Writer.TryWrite(entry);
+                    }
+    
+                    dropSummary[log.LogInfo.LogLevel]++;
+                    lock (lockObject) didDrop = true;
+                }
+                else
+                {
+                    bool tryWrite;
+                    lock (lockObject) tryWrite = channel.Writer.TryWrite(log);
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: explain this lock?

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 6]
```diff
@@ -1,62 +1,212 @@
...
+                bool? spamming = PostsTimesCheck();
+                if (spamming.HasValue)
+                    lock (lockObject)
+                        isSpamming = spamming.Value;
+
+                return isSpamming;
+            }
+        }
+
+        private bool? PostsTimesCheck()
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs:182]: maybe document it a bit (bool -> null means don't change current status, etc.)

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 7]
```diff
@@ -190,7 +343,98 @@ await Task.Delay(sleepTime, cancellationTokenSource.Token)
...
+
+                if (!didDrop) continue;
+
+                if (dropSummary[LogLevel.Trace]       == 0 &&
+                    dropSummary[LogLevel.Debug]       == 0 &&
+                    dropSummary[LogLevel.Information] == 0 &&
+                    dropSummary[LogLevel.Warning]     == 0 &&
+                    dropSummary[LogLevel.Error]       == 0 &&
+                    dropSummary[LogLevel.Critical]    == 0 &&
+                    dropSummary[LogLevel.None]        == 0)
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: can be removed

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs [hunk 8]
```diff
@@ -190,7 +343,98 @@ await Task.Delay(sleepTime, cancellationTokenSource.Token)
...
+                        dropSummary[key] = 0;
+                    }
+                }
+                catch
+                {
+                    // ignored
+                }
+                finally
+                {
+                    didDrop = false;
```
Comment (thebne) [inline src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: I don't really understand what `didDrop` represents (so maybe rename)

**Review by thebne — CHANGES_REQUESTED** [review]
explain the lock solution and address the rest

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 1 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs:447]: Done on: e9ffc95 - Task.WhenAll dispose

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 2 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: This is just the code formatter I use, this is too much to revert and redo the entire PR just for blank lines

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 3 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: I Suggest:

50 / 1sec In
40 / 1sec Out (stop dropping logs after spam rate reaches the about condition)

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 4 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: 79b11fa - remove lock comment and braces

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 5 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: Changed it, see dab4f1f - lock entire channel queue and count q

Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: It looked like the writer on a channel was not concurrent, so writes where overriding each other sometimes

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 6 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs:182]: a700c1c - minor document function

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 7 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: a700c1c - minor document function

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 8 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: It is for the use of `SummaryWriteLoop`
Also added lock for the read and write
960b2ef - lock didDrop bool

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 3 above)
Comment (thebne) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: the thresholds should be bigger to accommodate for spikes in log writes (scene load etc.).
and we need to assume 45fps.
so maybe 200/5sec instead? (over 40 messages for 5 sec)

also now that I'm thinking about it, we need a way to disable it in editor.

can you test our production code (not a sample scene) with these scenarios?

1. combat, a script throws an error every frame
2. combat, 3 scripts throw an error every frame
3. main menu, a script throws an error every frame

File: src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs (hunk 2 above)
Comment (kfird214) [reply src/ZLogger.Unity/Assets/Scripts/ZLogger/AsyncStreamLineMessageWriter.cs]: d75b89f - Remove spaces

**Pull request merged**